AX: Crash when navigating character-by-character via VoiceOver in text with multi-byte glyphs (e.g. emojis)

twilco opened the pull request:
#### 9a59ea98b0a6cb5746f0898a18118dba56b57c1a
<pre>
AX: Crash when navigating character-by-character via VoiceOver in text with multi-byte glyphs (e.g. emojis)
<a href="https://bugs.webkit.org/show_bug.cgi?id=293753">https://bugs.webkit.org/show_bug.cgi?id=293753</a>
<a href="https://rdar.apple.com/152255108">rdar://152255108</a>

Reviewed by Vitor Roriz.

In AccessibilityRenderObject::textRuns, when we build text runs for use off the main-thread, part of that process
is measuring and storing character widths. It&apos;s important that the character widths we cache match the length of the
string. This matters particularly when considering multi-byte glyphs like certain emojis, which can expand the length
of the string multiple times (e.g. some emojis are 10 characters long!) but only take up one position in the GlyphBuffer.

To handle this, we padded our cached character widths with zeros based on the result of GlyphBuffer::uncheckedStringOffsetAt,
which should tell us how big the gaps created by multi-byte glyphs are. Unfortunately, ComplexTextController did not
compute this value correctly when building the glyph buffer. It only added the glyph index as the string offset, and
the glyph index is only relevant in the context of a single ComplexTextRun, not the overall string. This is easily
fixed by also adding the ComplexTextRun::stringLocation to get the real string offset for the glyph. This fixes the
crash by ensuring we pad our character widths vector as intended, thus avoiding an out-of-bounds access assert.

This commit also fixes an issue where bounding boxes for text containing multi-byte glyphs became completely wrong.
This happened because when we converted an NSRange (a location, and length based on the whole length of the string,
including sub-characters of a multi-byte glyph), we walked over the multi-byte glyphs as one &quot;location&quot;, meaning
we would generate a text marker with the wrong offset.

Fix this by passing a new parameter to AXTextMarker::findMarker that forces walks through sub-characters of multi-byte
glyphs, rather than walking over them atomically. We&apos;ll need to use this in any place an accessibility API takes an
NSRange.

* LayoutTests/accessibility/mac/bounds-for-range-multibyte-glyphs-expected.txt: Added.
* LayoutTests/accessibility/mac/bounds-for-range-multibyte-glyphs.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::nextMarkerFromOffset const):
(WebCore::AXTextMarker::findMarker const):
* Source/WebCore/accessibility/AXTextMarker.h:
* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::localRect const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(computeTextBoundsForRange):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::enclosingGlyphBoundsForTextRun):

Canonical link: <a href="https://commits.webkit.org/295703@main">https://commits.webkit.org/295703@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8bbd076063d82f0281ff3cf9310fc8f62d9bf3c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105652 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25403 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15796 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110849 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56247 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107693 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25875 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33906 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80238 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108658 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20233 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95351 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60544 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19970 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13440 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55687 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89682 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13483 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113698 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32792 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24180 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89320 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33156 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91574 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88979 "Found 101 new API test failures: /TestWebKit:WebKit.PageLoadDidChangeLocationWithinPage, /TestWebKit:WebKit.AboutBlankLoad, /TestWebKit:WebKit.PreventEmptyUserAgent, /TestWebKit:WebKit.PrivateBrowsingPushStateNoHistoryCallback, /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/content-type, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/pointer-lock-permission-request, /TestWebKit:WebKit.HitTestResultNodeHandle, /TestWebKit:WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginShouldNotBeDispatchedForAlreadyFocusedField, /TestWebKit:WebKit.Find, /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebProcessExtension/dom-input-element-is-user-edited ... (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22736 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33852 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11658 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28252 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32717 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38109 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32463 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35812 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34061 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -2870,8 +2870,8 @@ static NSRect computeTextBoundsForRange(NSRange range, const AXCoreObject& backi
     if (AXObjectCache::useAXThreadTextApis()) {
-        auto markerToLocation = AXTextMarker { backingObject, 0 }.nextMarkerFromOffset(range.location);
-        auto markerToRangeEnd = markerToLocation.nextMarkerFromOffset(range.length);
+        auto markerToLocation = AXTextMarker { backingObject, 0 }.nextMarkerFromOffset(range.location, ForceSingleOffsetMovement::Yes);
```
Why ForceSingleOffsetMovement::Yes here? Don't we always want the next glyph?

twilco replied:
Take this example:

Abc⁉️def🧑‍🧑‍🧒‍🧒ghi👫Abc

String length is 27, but number of rendered glyphs is 15.

VoiceOver requests `NSAccessibilityBoundsForRangeParameterizedAttribute` with `NSRange { location: 25, length: 1 }`. This would be character "b"

To compute the start marker (markerToLocation in the PR), we call `AXTextMarker { backingObject, 0 }.nextMarkerFromOffset(range.location, ForceSingleOffsetMovement::Yes)`.

This walks forward 25 times to try to produce the right marker. But there are only 15 glyphs. We walk past "A", "b", and "c" just fine, decrementing counter from the initial 25, to 22.

Now we get to ⁉️. This is a 2 byte / character glyph. If we walk over it atomically, we move two places forward in the string, but only decrement counter once. Now we're at counter 21.

Subtract another 3 for ASCII "def". Counter = 18

Next is 🧑‍🧑‍🧒‍🧒. This is 10 bytes / characters (or 11? can't remember). We walk it atomically, meaning our counter is 17. So despite moving forward 10 / 11 characters in the string, we're only decrementing our counter once.

"ghi" -> counter 14

👫 -> two bytes I think, counter 13 because we walked it atomically

"Abc" -> counter 10

So we were asked to provide a text marker for location 25, but we're all the way at the end of the string, and have 10 more decrements left to go. This means we produce the wrong marker (one with offset 27)

vitorroriz commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -624,7 +624,7 @@ void ComplexTextController::advance(unsigned offset, GlyphBuffer* glyphBuffer, G
                 }
                 setHeight(paintAdvance, -height(paintAdvance)); // Increasing y points down
-                glyphBuffer->add(m_adjustedGlyphs[glyphIndexIntoComplexTextController], complexTextRun.font(), paintAdvance, complexTextRun.indexAt(m_glyphInCurrentRun), FloatPoint(textAutoSpaceSpacing, 0));
+                glyphBuffer->add(m_adjustedGlyphs[glyphIndexIntoComplexTextController], complexTextRun.font(), paintAdvance, complexTextRun.indexAt(m_glyphInCurrentRun) + complexTextRun.stringLocation(), FloatPoint(textAutoSpaceSpacing, 0));
```
I think this is reasonable, I was having a look if we would need something like that on `WidthIterator` as well, but I think not. 

`ComplexTextController` splits the original string into substrings for each `ComplexTextRun` when font transition happens. Currently the offset registered in the glyphbuffer is in relation to this ComplexTextRun substring. I think it makes sense to change that as the owner of `ComplexTextController` has access to the glyph buffer and to the original string but not to the individual ComplexTextRuns

twilco committed:
AX: Crash when navigating character-by-character via VoiceOver in text with multi-byte glyphs (e.g. emojis)
https://bugs.webkit.org/show_bug.cgi?id=293753
rdar://152255108

Reviewed by Vitor Roriz.

In AccessibilityRenderObject::textRuns, when we build text runs for use off the main-thread, part of that process
is measuring and storing character widths. It's important that the character widths we cache match the length of the
string. This matters particularly when considering multi-byte glyphs like certain emojis, which can expand the length
of the string multiple times (e.g. some emojis are 10 characters long!) but only take up one position in the GlyphBuffer.

To handle this, we padded our cached character widths with zeros based on the result of GlyphBuffer::uncheckedStringOffsetAt,
which should tell us how big the gaps created by multi-byte glyphs are. Unfortunately, ComplexTextController did not
compute this value correctly when building the glyph buffer. It only added the glyph index as the string offset, and
the glyph index is only relevant in the context of a single ComplexTextRun, not the overall string. This is easily
fixed by also adding the ComplexTextRun::stringLocation to get the real string offset for the glyph. This fixes the
crash by ensuring we pad our character widths vector as intended, thus avoiding an out-of-bounds access assert.

This commit also fixes an issue where bounding boxes for text containing multi-byte glyphs became completely wrong.
This happened because when we converted an NSRange (a location, and length based on the whole length of the string,
including sub-characters of a multi-byte glyph), we walked over the multi-byte glyphs as one "location", meaning
we would generate a text marker with the wrong offset.

Fix this by passing a new parameter to AXTextMarker::findMarker that forces walks through sub-characters of multi-byte
glyphs, rather than walking over them atomically. We'll need to use this in any place an accessibility API takes an
NSRange.

* LayoutTests/accessibility/mac/bounds-for-range-multibyte-glyphs-expected.txt: Added.
* LayoutTests/accessibility/mac/bounds-for-range-multibyte-glyphs.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::nextMarkerFromOffset const):
(WebCore::AXTextMarker::findMarker const):
* Source/WebCore/accessibility/AXTextMarker.h:
* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::localRect const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(computeTextBoundsForRange):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::enclosingGlyphBoundsForTextRun):

Canonical link: https://commits.webkit.org/295703@main
